### PR TITLE
Heedls 569 edit complete by date

### PR DESCRIPTION
**Pull request by AlexJacksonDS**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-569

### Description
Added a new Edit Complete By Date page linked from the View Delegate course cards and the Delegate Progress page

### Screenshots
image
image
image

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data/Services/ProgressService.cs [hunk 1]
```diff
@@ -48,5 +50,17 @@ public void UpdateSupervisor(int progressId, int? newSupervisorId)
...
+        public void UpdateCompleteByDate(int progressId, DateTime? completeByDate)
+        {
+            var courseInfo = courseDataService.GetDelegateCourseInfoByProgressId(progressId);
+
+            if (courseInfo == null)
+            {
+                throw new ProgressNotFoundException($"No progress record found for ProgressID {progressId}");
+            }
+
+            courseDataService.SetCompleteByDate(progressId, courseInfo.DelegateId,completeByDate);
```
Comment (wilkinsm) [inline DigitalLearningSolutions.Data/Services/ProgressService.cs:63]: There's a missing space before `completeByDate` param.

**Review by wilkinsm — APPROVED** [review]
Nice one Alex, just one minor comment.

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateProgressController.cs [hunk 2]
```diff
@@ -84,12 +85,58 @@ DelegateProgressAccessRoute accessedVia
...
+        [Route("EditCompleteByDate")]
+        public IActionResult EditCompleteByDate(
+            EditCompleteByDateFormData formData,
+            int progressId,
+            DelegateProgressAccessRoute accessedVia
+        )
+        {
+            if (!ModelState.IsValid)
+            {
+                var model = new EditCompleteByDateViewModel(formData, progressId, accessedVia);
```
Comment (rhimcg) [inline DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateProgressController.cs:117]: When there is an error the course name disappears so it just shows 'Edit complete date for'

Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateProgressController.cs:117]: Resolved this for this this ticket.

**Review by SteveJacksonSoft — CHANGES_REQUESTED** [review]

**Review by SteveJacksonSoft — APPROVED** [review]
Lovely, merge away.

**Pull request merged**